src/daft-core/src/array/ops/cast.rs

```diff
@@ -1527,6 +1609,308 @@ impl TensorArray {
     }
 }
 
+macro_rules! implement_cast_for_dense_with_inner_dtype {
+    ($dtype:ty, $array_type:ty, $n_values:ident, $non_zero_indices_array:ident, $non_zero_values_array:ident, $offsets:ident) => {{
+        let mut values = vec![0 as $dtype; $n_values];
+        let validity = $non_zero_values_array.validity();
+        for (i, indices) in $non_zero_indices_array.into_iter().enumerate() {
+            let is_valid = validity.map_or(true, |v| v.get_bit(i));
+            if !is_valid {
+                continue;
+            }
+            for j in 0..indices.unwrap().len() {
+                let index = $non_zero_indices_array
+                    .get(i)
+                    .unwrap()
+                    .u64()
+                    .unwrap()
+                    .as_arrow()
+                    .value(j) as usize;
+                let list_start_offset = $offsets.start_end(i).0;
+                values[list_start_offset + index] = $non_zero_values_array
+                    .get(i)
+                    .unwrap()
+                    .downcast::<$array_type>()
+                    .unwrap()
+                    .as_arrow()
+                    .value(j);
+            }
+        }
+        Box::new(arrow2::array::PrimitiveArray::from_vec(values))
+    }};
+}
+
+fn cast_sparse_to_dense_for_inner_dtype(
+    inner_dtype: &DataType,
+    n_values: usize,
+    non_zero_indices_array: &ListArray,
+    non_zero_values_array: &ListArray,
+    offsets: &Offsets<i64>,
+) -> DaftResult<Box<dyn arrow2::array::Array>> {
+    let item: Box<dyn arrow2::array::Array> = match inner_dtype {
```

Comment:
for these types we have the `with_match_numeric_daft_types` macro.
see https://github.com/Eventual-Inc/Daft/blob/6fe408c5badbd8621d4714baff2337aecaca74c6/src/daft-core/src/series/ops/between.rs#L36 as an example.